Refonte du système de build front (gulp)

sandhose opened the pull request:
| Q | R |
| --- | --- |
| Correction de bugs ? | oui |
| Nouvelle Fonctionnalité ? | oui |
| Tickets concernés | - |

Il s'agit d'une refonte complète du Gulpfile, en vue du futur stack JS, ce qui implique quelques changements:
- suppression de bower, on utilise npm pour toutes les dépendances 
- suppression du polyfill `css3-media-queries`, qui servait pour IE <= 8. On ne supporte de toutes façons pas ces versions.
- remplacement de gulp.spritesmith par css-sprite, qui est plus rapide, gère le rétina, et est plus stable
- intégration de gulp-notify, pour afficher les erreurs sous forme de notifications de bureau, si disponible
- suppression de gulp-newer, qui évitait de re-build dans certains cas, mais posait plus problème qu'autre chose

Le Gulpfile est maintenant plus clair, avec des commentaires, plus stable, et plus rapide (5.5s contre 8s auparavant sur ma bécane)

Avec le changement dans la génération du sprite, j'en ai profité pour correctement gérer le rétina: auparavant, les images étaient forcément chargées en @2x. Maintenant, cela dépend de la densité de pixel de la machine. C'est une feature qui peut potentiellement casser à certains endroits, il faudrait tester tout ce qui est icônes/images...
## Migration

Y'a pas grand chose à faire, à une exception: histoire de correctement repartir de zéro, il faudrait lancer un `npm run gulp -- clean` pour nettoyer l'environnement de travail. Sinon, c'est la procédure normale: `npm install`, et c'est bon. A noter aussi que quelques dépendances ont été virées, et qu'elles peuvent être désinstallées via `npm prune`.
## Notes QA

J'ai envie de dire, tester l'ensemble du site, mais c'est compliqué... Il faut d'abord re-build le front, avec `npm run gulp -- build`. Il faudrait surtout tester les images, et tout ce qui concerne le javascript (éditeur, menu mobile, autocomplétion, ...)

sandhose committed:
Refonte du système de build (gulp) :sparkles:

sandhose committed:
Suppression de bower :ok_hand:

GerardPaligot commented:
Les emojis dans les commentaires de commit, c'est mal !

sandhose commented:
Je savais que tu apprécierais :v:

sandhose committed:
Fix travis build + jshint :thumbsup:

gustavi commented on .gitignore:
```diff
@@ -19,11 +19,11 @@ base.db
-/assets/.sass-css
-/assets/bower_components
+/npm-debug.log
+# Générés par gulp
```
Commentaire à mettre en anglais

Situphen commented on package.json:
```diff
@@ -23,26 +20,28 @@
   },
   "homepage": "https://github.com/zestedesavoir/zds-site",
   "dependencies": {
```
Tu as oublié de séparer les dépendances des dépendances de développement !

sandhose committed:
Ajout des scripts npm "build" et "clean"
+ passage de certaines dépendances en dev
`npm run build` et `npm run clean` pour build/clean avec npm < 2.x
Evite de devoir installer gulp globalement, ou de mettre à jour npm

sandhose commented:
J'ai séparé les deps dev et les deps nécessaire au build ; si on est dans un environnement de production, l'on peut installer que ce qui est nécessaire via `npm install --production`.

J'en ai profité pour rajouter les scripts npm `clean` et `build`. Autrement dit, les personnes qui ont une version de npm < 2.x peuvent faire `npm run build` pour build le front.

GerardPaligot commented:
Juste pour signaler que la mise à jour du front s'est bien passée sur OS X grâce aux commandes :

```
sudo npm install -g npm # Il faut absolument être dans une version 2.x et supérieur.
npm install
npm run build
```

Situphen commented:
Je viens de tester, et sur mon ordinateur j'ai ça :

> [11/Jan/2015 17:30:33] "GET /static/images/sprite.png HTTP/1.1" 200 52015
> [11/Jan/2015 17:30:33] "GET /static/images/logo.png HTTP/1.1" 200 4769
> [11/Jan/2015 17:30:38] "GET /static/images/logo-mobile@2x.png HTTP/1.1" 200 6616

Ce qui veut dire que il n'y a que le logo-mobile qui se met en mode retina...

Situphen commented:
Aussi, je sais pas si ça vaut le coup de mettre les scripts "build" et "clean" dans le package.json pour les personnes qui ont la flemme de lancer la commande `sudo npm install -g npm`...

sandhose commented:
Bah, si jamais on change pour X raison de système de build, ils pourront toujours faire npm run [build|clean], et ca marchera comme il faut :-°

Situphen commented:
Bah du coup, met aussi les scripts "watch" et "test"...

sandhose commented:
test, c'est déjà le cas (convention de npm), et le watch est de toutes façons lancé avec npm run gulp sans arguments... Après, c'est une tâche vraiment de dev, et les devs front devraient avoir un environnement correct... C'est juste des shortcuts, on va pas s'emmerder avec ça maintenant ^^

sandhose committed:
Logo non-rétina pour le layout mobile

sandhose commented:
> Ce qui veut dire que il n'y a que le logo-mobile qui se met en mode retina...

C'est fixé!

pierre-24 commented:
Rapport de QA : 

Ok pour moi. Par acquis de conscience, j'ai testé sur navigateur et mobile (j'ai pas un rétina, ceci dit, mais, c'pas grave, j'ai un écran haute densité sur mon pc).

Ping @SpaceFox : bon pour moi :)

Situphen commented:
Ping @SpaceFox @sandhose : **Ne pas MEpP tout de suite !**

Vu que la tâche "pack" n'existe plus, il faut mettre à jour la script de déploiement.

Il faut aussi lancer un `npm prune` pour enlever les anciennes dépendances. Soit je met ça dans le "update.md", soit carrément dans le script de déploiement pour que ce soit fait à chaque fois. Sandhose, un avis ?

Du coup, soit vous faites une PR ce soir ou demain matin. Soit je met à jour ma PR #1926 demain après-midi. Voilà :)

sandhose commented:
Tu peux modifier ta PR, je pense ; d'ailleurs, tu peux rajouter l'option "--production" au npm update, pour installer que ce qui est nécessaire

Situphen commented:
@sandhose : et pour ce qui est du `npm prune --production`, je le met que pour cette fois-ci dans le "update.md" ou dans le script de déploiement pour faire du nettoyage à chaque fois ? (Je prône la deuxième solution.)

sandhose commented:
Ouais, ca peut se faire, dans le script de déploiement!

SpaceFox commented:
@Situphen  dans la doc `update.md`tu dis :

> Lancer la commande npm -v et voir le résultat.

Sauf qu'en préprod j'obtiens :

> ```
> zds@preprod:~$ npm -v
> -su: /usr/local/bin/npm : /usr/local/bin/node : mauvais interpréteur: Aucun fichier ou dossier de ce type
> ```

Je fais quoi ?

Situphen commented:
Euh, essayes :

``` shell
source /usr/local/nvm/nvm.sh
npm -v
```

SpaceFox commented:
Il me réponds "2.1.12", ce qui m'a l'air d'être OK.

On fait ce "source" dans le script aussi. C'est une bonne pratique ?

2015-01-17 20:51 GMT+01:00 Situphen notifications@github.com:

> Euh, essayes :
> 
> source /usr/local/nvm/nvm.sh
> npm -v
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zestedesavoir/zds-site/pull/2042#issuecomment-70381603
> .

Situphen commented:
C'est effectivement OK.

Je pense que oui. D'après ce que je sais, c'est un peu l'équivalent de virtualenv mais pour Node.JS, c'est-à-dire que ça permet d'avoir plusieurs versions différentes de Node.JS sur le même OS.

Eskimon commented:
`source` c'est pas une commande unix qui dit "garde moi dans le systeme les variables défini dans ce scirpt" ? (car sinon c'est derniere ont une portee "script" ou un truc comme ca)

gustavi commented:
SO : « source is a bash shell built-in command that executes the content of the file passed as argument, in the current shell. It has a synonym in '.' (period). »

http://ss64.com/bash/source.html

SpaceFox commented:
@sandhose  y'a moyen de mettre la doc d'install du front à jour ?

Parce que là, ce que j'ai dû faire en réinstallant ma VM de dev c'est ça :

<SpaceFox> \o/ victoire j'arrive à build le front (OK, sous Linux)
<SpaceFox> alors
<SpaceFox> 1. Installer NPM https://github.com/joyent/node/wiki/Installing-Node.js-via-package-manager
<SpaceFox> 2. Mettre à jour npm : https://github.com/zestedesavoir/zds-site/blob/dev/update.md#npm
<SpaceFox> oui, l'étape 1 est l'install officielle mais n'installe PAS la dernière verison
<SpaceFox> *version
<SpaceFox> 3. installer les outils front et build le front selon les instructions de QA ici : https://github.com/zestedesavoir/zds-site/pull/2042
<SpaceFox> et là, "chez moi ça marche"

Situphen commented:
@SpaceFox : tu parles bien de cette doc là https://github.com/zestedesavoir/zds-site/blob/dev/doc/gulp.md ? Normalement les instructions d'installation de npm de la doc sont à jour.

@sandhose : Je viens de voir quelques coquilles sur la doc, je peux te faire une PR ?

GerardPaligot commented:
@Situphen La doc mentionne l'usage de gulp alors qu'il n'est plus nécessaire.

SpaceFox commented:
J'ai l'impression qu'il n'y a pas de lien qui mène à cette doc. Il faudrait vérifier qu'en suivant les liens "installer ZdS" on arrive dessus.

Situphen commented:
Gulp est toujours nécessaire ! Il y a maintenant des raccourcis (`npm run build` à la place de `npm run gulp -- build` et `npm run clean` pour `npm run gulp -- clean`), complètement inutiles d'après moi, mais Gulp reste nécessaire.

SpaceFox commented:
Ah ouais, quand même...

Situphen commented:
@SpaceFox : c'est vrai que c'est fouillis, je vais faire une PR pour remettre plus de cohérence entre les docs !

sandhose commented:
@situphen : tu peux faire une Pr direct sur le répo principal, nan ?

@Andro: alors on utilise toujours Gulp, sauf qu'au lieu d'utiliser l'outil
"global" (qui doit être dans le PATH, toussa toussa...), on utilise Gulp,
mais installé localement,  dans node_modules/.bin/gulp, et on a rajouté un
script npm pour "simplifier" la commande (en réalité, "npm run gulp - -",
c'est juste un alias vers "node_modules/.bin/gulp"...)

Pour ce qui est de npm qui est pas à la dernière version, c'est la faute à
joyent, et à cause du modèle de gouvernance actuel de node, trop contrôlé
par cette société, ce qui a d'ailleurs mené à un fork de node (qui lui
embarque la dernière version de npm dans l'installer)... Il parait que ce
fork risque d'être merge un jour ou l'autre, donc peut-être qu'un jour, on
sera plus obligé de mettre à jour npm à la première Install ^^

Bon bah situphen a l'air chaud pour revoir tout ça, je le laisse faire :p
(faut juste penser à me ping quand tu fais ta Pr ;) )

Le mer. 21 janv. 2015 23:27, SpaceFox notifications@github.com a écrit :
